packages/expo-dev-launcher/android/src/main/java/expo/modules/devlauncher/modules/DevLauncherInternalModule.kt

```diff
@@ -0,0 +1,243 @@
+  private val controller: DevLauncherControllerInterface by inject()
+  private val intentRegistry: DevLauncherIntentRegistryInterface by inject()
+  private val installationIDHelper: DevLauncherInstallationIDHelper by inject()
+
+  override fun definition() = ModuleDefinition {
+    Name("ExpoDevLauncherInternal")
+    Events(ON_NEW_DEEP_LINK_EVENT)
+
+    OnCreate {
+      if (wasInitialized()) {
+        intentRegistry.subscribe(this@DevLauncherInternalModule::onNewPendingIntent)
+      }
+    }
+
+    OnDestroy {
+      if (wasInitialized()) {
+        intentRegistry.unsubscribe(this@DevLauncherInternalModule::onNewPendingIntent)
+      }
+    }
+
+    Constants {
+      val isRunningOnEmulator = EmulatorUtilities.isRunningOnEmulator()
+
+      return@Constants mapOf(
+        "installationID" to installationIDHelper.getOrCreateInstallationID(context),
+        "isDevice" to !isRunningOnEmulator,
+        "updatesConfig" to getUpdatesConfig()
+      )
+    }
+
+    AsyncFunction("loadUpdate") { url: String, projectUrlString: String?, promise: Promise ->
+      controller.coroutineScope.launch {
+        try {
+          val appUrl = sanitizeUrlString(url)
+          var projectUrl: Uri? = null
+          if (projectUrlString != null) {
+            projectUrl = sanitizeUrlString(projectUrlString)
+          }
+          controller.loadApp(appUrl, projectUrl)
+        } catch (e: Exception) {
+          promise.reject("ERR_DEV_LAUNCHER_CANNOT_LOAD_APP", e.message, e)
+          return@launch
+        }
+        promise.resolve(null)
+      }
+    }
+
+    AsyncFunction("loadApp") { url: String, promise: Promise ->
+      controller.coroutineScope.launch {
+        try {
+          val parsedUrl = sanitizeUrlString(url)
+          controller.loadApp(parsedUrl)
+        } catch (e: Exception) {
+          promise.reject("ERR_DEV_LAUNCHER_CANNOT_LOAD_APP", e.message, e)
+          return@launch
+        }
+        promise.resolve(null)
+      }
+    }
+
+    AsyncFunction<List<Bundle?>>("getRecentlyOpenedApps") {
+      val apps: MutableList<Bundle> = ArrayList()
+
+      for (recentlyOpenedApp in controller.getRecentlyOpenedApps()) {
+        val app = Bundle()
+
+        app.putDouble("timestamp", recentlyOpenedApp.timestamp.toDouble())
+        app.putString("name", recentlyOpenedApp.name)
+        app.putString("url", recentlyOpenedApp.url)
+        app.putBoolean("isEASUpdate", recentlyOpenedApp.isEASUpdate == true)
+
+        if (recentlyOpenedApp.isEASUpdate == true) {
+          app.putString("updateMessage", recentlyOpenedApp.updateMessage)
+          app.putString("branchName", recentlyOpenedApp.branchName)
+        }
+
+
+        apps.add(app)
+      }
+      return@AsyncFunction apps
+    }
+
+    AsyncFunction("clearRecentlyOpenedApps") {
+      controller.clearRecentlyOpenedApps()
+    }
+
+    AsyncFunction("getPendingDeepLink") { promise: Promise ->
+      intentRegistry.intent?.data?.let {
+        promise.resolve(it.toString())
```

Comment:
Should we return early so we don't resolve the promise twice?

Reply:
sure, updated